### DISCUSSION TITLE
SQL kernel backed by sql.js

### DISCUSSION THREAD

**Discussion by jtpio** [jupyterlite/jupyterlite#96] [Ideas]
This was initially mentioned by @bollwyvl on Gitter: https://gitter.im/jupyterlab/jupyterlab?at=60a0202f5ca66330b0456558

Opening a discussion here so it's easier to track.

It would indeed be cool to have a SQL kernel backed by `sql.js`!

Maybe a good candidate for a serverlite federated extension?

References:

- https://phiresky.github.io/blog/2021/hosting-sqlite-databases-on-github-pages/
- https://github.com/sql-js/sql.js/

**Comment by bollwyvl** [comment]
I've done a lot of the scut work on this over on [janki](https://github.com/gt-coar/janki/pull/7), and might or might not get continued funding there.

The sqlite.js plumbing piece is pretty much standalone, and delegates stuff into a hand-made `comlink`... it loses a few rad features, like custom js functions in your sql :exploding_head: but could be repurposed with an interpreter-style thing.

The other side of that is `libarchive` (the objective of janki is updating a sqlite database inside a zipfile which can also contain images, audio, movies :scream: ) which has a _lot_ of potential uses... like self-hosting `jupyter lite build`.

while pyodide also supports these features, it's 1/10th of the size to just get to queries (before one starts loading a database).

**Comment by twoxfh** [comment]
Any reason we should not use the existing one in Python? If its used, it might be worth updating the version.

![image](https://user-images.githubusercontent.com/30451770/134358443-820dcfd9-faad-4dcb-a17f-5856661b208e.png)

**Reply by bollwyvl** [reply]
With our goal of eventually different _lite_ experiences, sqlite is a good good choice:
- has `lite` in the name: jokes aside, it shares some of the design goals
- `sqlite.js` is ~1.5mb, while `pyodide` (well, `pyolite`, as we use it) _starts_ at about 20mb with an empty cache (which we also need to improve :blush:)
- the build chain for getting the `sqlite` version number bumped in `pyodide` is presently... long, and the release cadence rather stately (unless significant bugs are found)

I'm imagining, with a _few_ built-in renderers (e.g. `vega`, `datagrid`) a sqlite kernel could be both much more lightweight, and probably more responsive/faster, and lay the groundwork for other standards-based not-quite-programming languages (e.g. SPARQL).

Further, the crazy ["sqlite-virtual-file-system-over-http"](https://phiresky.github.io/blog/2021/hosting-sqlite-databases-on-github-pages/) would allow _well-indexed_ datasets of gigabytes to be queried and visualized at interactive speed, only downloading the data actually used... integrating that into the python side of the world would be a lot more complicated.

**Reply by twoxfh** [reply]
My interpretation of lite is it needs to be available offline which may not gel with sqlite over http. Absurd-sql uses a similar approach but its contained within the browser storage. All interesting projects and possibilities.

**Comment by DerThorsten** [comment]
Hi, 
There is also some discussion on [xeus-sqlite](https://github.com/jupyter-xeus/xeus-sqlite/issues/118) to add a xeus based jupyterlite kernel. 
I am currently looking into this!
